File: src/Slightly.cpp
// Slightly.C

#include "Slightly.H"

#define EPSI (1e-25)

double Slightly::less(double x) {
  double epsi = EPSI;
  double x0 = x-epsi;
  while (x0==x) {
    epsi *= 10;
    x0 = x-epsi;
  }
  return x0;
}
    
double Slightly::more(double x) {
  double epsi = EPSI;
  double x0 = x+epsi;
  while (x0==x) {
    epsi *= 10;
    x0 = x+epsi;
  }
  return x0;
}
    

File: python/qplot_backend/slightly.py
# slightly.py - This file is part of QPlot

# QPlot - Publication quality 2D graphs with dual coordinate systems

import math


def slightly_less(x: float) -> float:
    """Return the largest representable float strictly less than x."""
    return math.nextafter(x, -math.inf)


def slightly_more(x: float) -> float:
    """Return the smallest representable float strictly greater than x."""
    return math.nextafter(x, math.inf)
